### PR TITLE
[SPARK-22165][SQL] Fixes type conflicts between double, long, decimals, dates and timestamps in partition column

### PR DESCRIPTION
**Pull request by HyukjinKwon**
## What changes were proposed in this pull request?

This PR proposes to add a rule that re-uses `TypeCoercion.findWiderCommonType` when resolving type conflicts in partition values.

Currently, this uses numeric precedence-like comparison; therefore, it looks introducing failures for type conflicts between timestamps, dates and decimals, please see:

```scala
private val upCastingOrder: Seq[DataType] =
  Seq(NullType, IntegerType, LongType, FloatType, DoubleType, StringType)
...
literals.map(_.dataType).maxBy(upCastingOrder.indexOf(_))
```

The codes below:

```scala
val df = Seq((1, "2015-01-01"), (2, "2016-01-01 00:00:00")).toDF("i", "ts")
df.write.format("parquet").partitionBy("ts").save("/tmp/foo")
spark.read.load("/tmp/foo").printSchema()

val df = Seq((1, "1"), (2, "1" * 30)).toDF("i", "decimal")
df.write.format("parquet").partitionBy("decimal").save("/tmp/bar")
spark.read.load("/tmp/bar").printSchema()
```

produces output as below:

**Before**

```
root
 |-- i: integer (nullable = true)
 |-- ts: date (nullable = true)

root
 |-- i: integer (nullable = true)
 |-- decimal: integer (nullable = true)
```

**After**

```
root
 |-- i: integer (nullable = true)
 |-- ts: timestamp (nullable = true)

root
 |-- i: integer (nullable = true)
 |-- decimal: decimal(30,0) (nullable = true)
```

### Type coercion table:

This PR proposes the type conflict resolusion as below:

**Before**

|InputA \ InputB|`NullType`|`IntegerType`|`LongType`|`DecimalType(38,0)`|`DoubleType`|`DateType`|`TimestampType`|`StringType`|
|------------------------|----------|----------|----------|----------|----------|----------|----------|----------|
|**`NullType`**|`StringType`|`IntegerType`|`LongType`|`StringType`|`DoubleType`|`StringType`|`StringType`|`StringType`|
|**`IntegerType`**|`IntegerType`|`IntegerType`|`LongType`|`IntegerType`|`DoubleType`|`IntegerType`|`IntegerType`|`StringType`|
|**`LongType`**|`LongType`|`LongType`|`LongType`|`LongType`|`DoubleType`|`LongType`|`LongType`|`StringType`|
|**`DecimalType(38,0)`**|`StringType`|`IntegerType`|`LongType`|`DecimalType(38,0)`|`DoubleType`|`DecimalType(38,0)`|`DecimalType(38,0)`|`StringType`|
|**`DoubleType`**|`DoubleType`|`DoubleType`|`DoubleType`|`DoubleType`|`DoubleType`|`DoubleType`|`DoubleType`|`StringType`|
|**`DateType`**|`StringType`|`IntegerType`|`LongType`|`DateType`|`DoubleType`|`DateType`|`DateType`|`StringType`|
|**`TimestampType`**|`StringType`|`IntegerType`|`LongType`|`TimestampType`|`DoubleType`|`TimestampType`|`TimestampType`|`StringType`|
|**`StringType`**|`StringType`|`StringType`|`StringType`|`StringType`|`StringType`|`StringType`|`StringType`|`StringType`|

**After**

|InputA \ InputB|`NullType`|`IntegerType`|`LongType`|`DecimalType(38,0)`|`DoubleType`|`DateType`|`TimestampType`|`StringType`|
|------------------------|----------|----------|----------|----------|----------|----------|----------|----------|
|**`NullType`**|`NullType`|`IntegerType`|`LongType`|`DecimalType(38,0)`|`DoubleType`|`DateType`|`TimestampType`|`StringType`|
|**`IntegerType`**|`IntegerType`|`IntegerType`|`LongType`|`DecimalType(38,0)`|`DoubleType`|`StringType`|`StringType`|`StringType`|
|**`LongType`**|`LongType`|`LongType`|`LongType`|`DecimalType(38,0)`|`StringType`|`StringType`|`StringType`|`StringType`|
|**`DecimalType(38,0)`**|`DecimalType(38,0)`|`DecimalType(38,0)`|`DecimalType(38,0)`|`DecimalType(38,0)`|`StringType`|`StringType`|`StringType`|`StringType`|
|**`DoubleType`**|`DoubleType`|`DoubleType`|`StringType`|`StringType`|`DoubleType`|`StringType`|`StringType`|`StringType`|
|**`DateType`**|`DateType`|`StringType`|`StringType`|`StringType`|`StringType`|`DateType`|`TimestampType`|`StringType`|
|**`TimestampType`**|`TimestampType`|`StringType`|`StringType`|`StringType`|`StringType`|`TimestampType`|`TimestampType`|`StringType`|
|**`StringType`**|`StringType`|`StringType`|`StringType`|`StringType`|`StringType`|`StringType`|`StringType`|`StringType`|

This was produced by:

```scala
  test("Print out chart") {
    val supportedTypes: Seq[DataType] = Seq(
      NullType, IntegerType, LongType, DecimalType(38, 0), DoubleType,
      DateType, TimestampType, StringType)

    // Old type conflict resolution:
    val upCastingOrder: Seq[DataType] =
      Seq(NullType, IntegerType, LongType, FloatType, DoubleType, StringType)
    def oldResolveTypeConflicts(dataTypes: Seq[DataType]): DataType = {
      val topType = dataTypes.maxBy(upCastingOrder.indexOf(_))
      if (topType == NullType) StringType else topType
    }
    println(s"|InputA \\ InputB|${supportedTypes.map(dt => s"`${dt.toString}`").mkString("|")}|")
    println(s"|------------------------|${supportedTypes.map(_ => "----------").mkString("|")}|")
    supportedTypes.foreach { inputA =>
      val types = supportedTypes.map(inputB => oldResolveTypeConflicts(Seq(inputA, inputB)))
      println(s"|**`$inputA`**|${types.map(dt => s"`${dt.toString}`").mkString("|")}|")
    }

    // New type conflict resolution:
    def newResolveTypeConflicts(dataTypes: Seq[DataType]): DataType = {
      dataTypes.fold[DataType](NullType)(findWiderTypeForPartitionColumn)
    }
    println(s"|InputA \\ InputB|${supportedTypes.map(dt => s"`${dt.toString}`").mkString("|")}|")
    println(s"|------------------------|${supportedTypes.map(_ => "----------").mkString("|")}|")
    supportedTypes.foreach { inputA =>
      val types = supportedTypes.map(inputB => newResolveTypeConflicts(Seq(inputA, inputB)))
      println(s"|**`$inputA`**|${types.map(dt => s"`${dt.toString}`").mkString("|")}|")
    }
  }
```

## How was this patch tested?

Unit tests added in `ParquetPartitionDiscoverySuite`.

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
cc @cloud-fan (I believe my similar PR was reviewed by you before), @ueshin and @squito.

**Comment by HyukjinKwon** [comment]
retest this please

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetPartitionDiscoverySuite.scala
Comment (squito) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetPartitionDiscoverySuite.scala]: just for my own understanding -- these added asserts are just to improve failure msgs, right?  They would have all been covered by the `assert(actualSpec === spec)` below anywyay, right?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetPartitionDiscoverySuite.scala
Comment (squito) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetPartitionDiscoverySuite.scala]: this passes even before the change, right?

as I mentioned on the other PR, I don't really understand why this works despite the issue you're fixing.  Regardless, seems like a good test to have.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala
Comment (squito) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala]: surprisingly, `TypeCoercion` is totally public, so this should probably be `private[sql]`.

**Review by squito — APPROVED** [review]
lgtm, just a small comment on scoping, though I'm not an expert on this area.  Left a couple of questions just for my own understanding.

Also, I'm curious why the tests are all going into `ParquetPartitionDiscoverySuite` -- this doesn't seem specific to parquet, and in fact I wonder if it will be different in parquet since the non-partition columns have schemas specified in the data.  I'm just surprised this isn't tested across more formats.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetPartitionDiscoverySuite.scala
Comment (HyukjinKwon) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetPartitionDiscoverySuite.scala]: Yes, it is. It was difficult for me to find which one was different.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala
Comment (HyukjinKwon) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala]: Actually, these were removed in [SPARK-16813](https://issues.apache.org/jira/browse/SPARK-16813) and https://github.com/apache/spark/pull/14418 :).

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetPartitionDiscoverySuite.scala
Comment (HyukjinKwon) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetPartitionDiscoverySuite.scala]: I think `2016-01-01 00:01:00` was placed when calling `resolveTypeConflicts`.  Let me try to make this test case not dependent on this.

**Comment by HyukjinKwon** [comment]
> why the tests are all going into `ParquetPartitionDiscoverySuite` -- this doesn't seem specific to parquet, and in fact I wonder if it will be different in parquet since the non-partition columns have schemas specified in the data. I'm just surprised this isn't tested across more formats.

Yea, I agree since this problem is not specific to Parquet. Here, such changes and test cases look added in this file so far and I simply just decided to follow it, rather than including changes here restructuing or moving the test cases, partly for the easy of backporting (we should backport this into branch-2.2 and 2.1) and partly to reduce reviewing cost.

**Comment by gatorsmile** [comment]
This PR introduces the behavior changes. We are unable to do this.

**Comment by gatorsmile** [comment]
Please ensure no behavior change is introduced when fixing such issues. Also cc @cloud-fan

**Comment by HyukjinKwon** [comment]
@gatorsmile, could you elaborate which behaviour changes you mean?

**Comment by HyukjinKwon** [comment]
Do you mean before / after in PR description? They are bugs to fix, aren't they?

**Comment by HyukjinKwon** [comment]
ping @gatorsmile

**Comment by HyukjinKwon** [comment]
ping?

**Comment by HyukjinKwon** [comment]
@cloud-fan, could you take a look when you have some time please?

**Comment by gatorsmile** [comment]
Will review it this weekend.

**Comment by HyukjinKwon** [comment]
Thank you so much @gatorsmile.

**Comment by HyukjinKwon** [comment]
Hi @gatorsmile, could you please review this when you have some time?

**Comment by gatorsmile** [comment]
Generally, the current type inference/coercion rules are messy and random. We have to seriously revisit our type coercion. 

After thinking it more, I think this change in this PR is pretty risky. It just introduces new type inference behaviors, although I do not like the previous one neither. These changes could easily cause new regression when our users upgrading their Spark versions. 

For making the migration more smooth, my general proposal is to introduce a conf for each one when we change something like this, if we believe this is a bug fix. Remove or deprecate the internal conf in the next release (or after a few releases) if nobody raises the issue after a major release (around half a year).

**Comment by gatorsmile** [comment]
Also cc @rxin @cloud-fan @sameeragarwal

**Comment by HyukjinKwon** [comment]
Maybe, what do you think about opening a discussion in the mailing list? If I understood correctly, committers have a different opinion on this (did I understand correctly?). That should deduplicate a discussion about it. I am willing to actively join in it.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningUtils.scala
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningUtils.scala]: I think it's a bug, the previous code didn't consider the case when input literal types are outside of the `upCastingOrder`, and just pick the first type as the final type.

However I'm not sure what's the expected behavior. We need to figure out what's the possible data types for partition columns, and how to merge them.

Comment (HyukjinKwon) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningUtils.scala]: Let me try to describe what this PR explicitly changes soon.

Comment (HyukjinKwon) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningUtils.scala]: I think we will have the input types for this `resolveTypeConflicts`:

```scala
Seq(
    NullType, IntegerType, LongType, DoubleType,
    *DecimalType(...), DateType, TimestampType, StringType)
```

*`DecimalType` only when it's bigger than `LongType`:

https://github.com/apache/spark/blob/04975a68b583a6175f93da52374108e5d4754d9a/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningUtils.scala#L384-L393



Because:

this particular `resolveTypeConflicts` seems being only called through:

https://github.com/apache/spark/blob/04975a68b583a6175f93da52374108e5d4754d9a/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningUtils.scala#L142

https://github.com/apache/spark/blob/04975a68b583a6175f93da52374108e5d4754d9a/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningUtils.scala#L337

https://github.com/apache/spark/blob/04975a68b583a6175f93da52374108e5d4754d9a/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningUtils.scala#L474

In the first call, I am seeing `pathsWithPartitionValues` is constructed by `partitionValues`, which is the output from `parsePartition`:

https://github.com/apache/spark/blob/04975a68b583a6175f93da52374108e5d4754d9a/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningUtils.scala#L108

which parses the input by `parsePartitionColumn`:

https://github.com/apache/spark/blob/04975a68b583a6175f93da52374108e5d4754d9a/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningUtils.scala#L209

which calls this `inferPartitionColumnValue`:


https://github.com/apache/spark/blob/04975a68b583a6175f93da52374108e5d4754d9a/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningUtils.scala#L384-L428

Comment (HyukjinKwon) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningUtils.scala]: So, for the types:

```scala
Seq(
    NullType, IntegerType, LongType, DoubleType,
    *DecimalType(...), DateType, TimestampType, StringType)
```


I produced a chart as below by this codes:

```scala
test("Print out chart") {
  val supportedTypes: Seq[DataType] = Seq(
    NullType, IntegerType, LongType, DoubleType,
    DecimalType(38, 0), DateType, TimestampType, StringType)

  val combinations = for {
    t1 <- supportedTypes
    t2 <- supportedTypes
  } yield Seq(t1,t2)

  // Old type conflict resolution:
  val upCastingOrder: Seq[DataType] =
    Seq(NullType, IntegerType, LongType, FloatType, DoubleType, StringType)

  def oldResolveTypeConflicts(dataTypes: Seq[DataType]): DataType = {
    val topType = dataTypes.maxBy(upCastingOrder.indexOf(_))
    if (topType == NullType) StringType else topType
  }

  // New type conflict resolution:
  def newResolveTypeConflicts(dataTypes: Seq[DataType]): DataType = {
    TypeCoercion.findWiderCommonType(dataTypes) match {
      case Some(NullType) => StringType
      case Some(dt: DataType) => dt
      case _ => StringType
    }
  }

  println("|Input types|Old output type|New output type|")
  println("|-----------|---------------|---------------|")
  combinations.foreach { pair =>
    val oldType = oldResolveTypeConflicts(pair)
    val newType = newResolveTypeConflicts(pair)

    if (oldType != newType) {
      println(s"|[`${pair(0)}`, `${pair(1)}`]|`$oldType`|`$newType`|")
    }
  }
}
```

So, looks this PR makes the changes in type resolution as below:

|Input types|Old output type|New output type|
|-----------|---------------|---------------|
|[`NullType`, `DecimalType(38,0)`]|`StringType`|`DecimalType(38,0)`|
|[`NullType`, `DateType`]|`StringType`|`DateType`|
|[`NullType`, `TimestampType`]|`StringType`|`TimestampType`|
|[`IntegerType`, `DecimalType(38,0)`]|`IntegerType`|`DecimalType(38,0)`|
|[`IntegerType`, `DateType`]|`IntegerType`|`StringType`|
|[`IntegerType`, `TimestampType`]|`IntegerType`|`StringType`|
|[`LongType`, `DecimalType(38,0)`]|`LongType`|`DecimalType(38,0)`|
|[`LongType`, `DateType`]|`LongType`|`StringType`|
|[`LongType`, `TimestampType`]|`LongType`|`StringType`|
|[`DoubleType`, `DateType`]|`DoubleType`|`StringType`|
|[`DoubleType`, `TimestampType`]|`DoubleType`|`StringType`|
|[`DecimalType(38,0)`, `NullType`]|`StringType`|`DecimalType(38,0)`|
|[`DecimalType(38,0)`, `IntegerType`]|`IntegerType`|`DecimalType(38,0)`|
|[`DecimalType(38,0)`, `LongType`]|`LongType`|`DecimalType(38,0)`|
|[`DecimalType(38,0)`, `DateType`]|`DecimalType(38,0)`|`StringType`|
|[`DecimalType(38,0)`, `TimestampType`]|`DecimalType(38,0)`|`StringType`|
|[`DateType`, `NullType`]|`StringType`|`DateType`|
|[`DateType`, `IntegerType`]|`IntegerType`|`StringType`|
|[`DateType`, `LongType`]|`LongType`|`StringType`|
|[`DateType`, `DoubleType`]|`DoubleType`|`StringType`|
|[`DateType`, `DecimalType(38,0)`]|`DateType`|`StringType`|
|[`DateType`, `TimestampType`]|`DateType`|`TimestampType`|
|[`TimestampType`, `NullType`]|`StringType`|`TimestampType`|
|[`TimestampType`, `IntegerType`]|`IntegerType`|`StringType`|
|[`TimestampType`, `LongType`]|`LongType`|`StringType`|
|[`TimestampType`, `DoubleType`]|`DoubleType`|`StringType`|
|[`TimestampType`, `DecimalType(38,0)`]|`TimestampType`|`StringType`|

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningUtils.scala]: I think the new behavior is much better. It seems like the previous behavior is just wrong, but it's very rare to see different data types in partition columns, and that's why no users open tickets for it yet.

cc @gatorsmile

Comment (gatorsmile) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningUtils.scala]: Since this changes the schema, we can't change it without a conf.

Comment (gatorsmile) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningUtils.scala]: We also need to document the behavior change in the doc.

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningUtils.scala]: Do we really wanna document the previous wrong behavior in the doc? An example is merging `TimestampType` and `DateType`, the result is non-deterministic, depends on which partition path gets parsed first. How do we document that?

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningUtils.scala]: BTW this is only used in partition discovery, I can't think of a problematic case for it. The only thing it can break is, users give Spark a data dir to do partition discovering, and users make an assumption on the inferred partition schema, but then we can argue that why users ask Spark to do partition discovery at the beginning?

Comment (gatorsmile) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningUtils.scala]: Both the values and the schemas could be changed. The external applications might be broken if the schema is different. 

The new behaviors are consistent with what we does for the other type coercion cases. However, implicit type casting and partition discovery are unstable. The other mature systems have clear/stable rules about it. Below is an example. https://docs.oracle.com/cloud/latest/db112/SQLRF/sql_elements002.htm#g195937

If each release introduces new behaviors, it becomes hard to use by the end users who have such expectation. Thus, my suggestion is to first stabilize our type coercion rules before addressing this. 

https://github.com/apache/spark/pull/18853 is the first PR attempt in this direction.

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningUtils.scala]: Partitioned columns are different from normal type coercion cases, they are literally all string type, and we are just trying to find a most reasonable type of them.

The previous behavior was there since the very beginning, which I think didn't go through a decent discussion. This is the first time we seriously design the type merging logic for partition discovery. I think it doesn't need to be blocked by the type coercion stabilization work, as they can diverge.

@HyukjinKwon can you send the proposal to dev list? I think we need more feedback, e.g. people may want more strict rules and have more cases to fallback to string.

Comment (HyukjinKwon) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningUtils.scala]: Let me send the proposal to dev soon tonight (KST)

**Comment by HyukjinKwon** [comment]
I have just made a table to check the diff easily:

**Before**:

|InputA \ InputB|`NullType`|`IntegerType`|`LongType`|`DecimalType(38,0)`|`DoubleType`|`DateType`|`TimestampType`|`StringType`|
|------------------------|----------|----------|----------|----------|----------|----------|----------|----------|
|**`NullType`**|`StringType`|||`StringType`||`StringType`|`StringType`||
|**`IntegerType`**||||`IntegerType`||`IntegerType`|`IntegerType`||
|**`LongType`**||||`LongType`|`DoubleType`|`LongType`|`LongType`||
|**`DecimalType(38,0)`**|`StringType`|`IntegerType`|`LongType`||`DoubleType`|`DecimalType(38,0)`|`DecimalType(38,0)`||
|**`DoubleType`**|||`DoubleType`|`DoubleType`||`DoubleType`|`DoubleType`||
|**`DateType`**|`StringType`|`IntegerType`|`LongType`|`DateType`|`DoubleType`||`DateType`||
|**`TimestampType`**|`StringType`|`IntegerType`|`LongType`|`TimestampType`|`DoubleType`||||
|**`StringType`**|||||||||

**After**:

|InputA \ InputB|`NullType`|`IntegerType`|`LongType`|`DecimalType(38,0)`|`DoubleType`|`DateType`|`TimestampType`|`StringType`|
|------------------------|----------|----------|----------|----------|----------|----------|----------|----------|
|**`NullType`**|`NullType`|||`DecimalType(38,0)`||`DateType`|`TimestampType`||
|**`IntegerType`**||||`DecimalType(38,0)`||`StringType`|`StringType`||
|**`LongType`**||||`DecimalType(38,0)`|`StringType`|`StringType`|`StringType`||
|**`DecimalType(38,0)`**|`DecimalType(38,0)`|`DecimalType(38,0)`|`DecimalType(38,0)`||`StringType`|`StringType`|`StringType`||
|**`DoubleType`**|||`StringType`|`StringType`||`StringType`|`StringType`||
|**`DateType`**|`DateType`|`StringType`|`StringType`|`StringType`|`StringType`||`TimestampType`||
|**`TimestampType`**|`TimestampType`|`StringType`|`StringType`|`StringType`|`StringType`||||
|**`StringType`**|||||||||

**Comment by HyukjinKwon** [comment]
retest this please

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningUtils.scala
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningUtils.scala]: we should also put this table in sql-programming-guide and sql migration guide.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningUtils.scala
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningUtils.scala]: nit: I think we can use `reduce`? `literals` should not be empty.

File: docs/sql-programming-guide.md
Comment (HyukjinKwon) [inline docs/sql-programming-guide.md]: Built doc shows as below:

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningUtils.scala
Comment (HyukjinKwon) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningUtils.scala]: Yup. I just address this one and the one above.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningUtils.scala
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningUtils.scala]: might be good to explain why we can only see `DecimaType(38, 0)`

**Comment by cloud-fan** [comment]
LGTM

**Comment by cloud-fan** [comment]
One thing we can do in follow-up: the sql programming guide does have a partition discovery section, but it's under the parquet section, we should move it a layer up and put the type casting table there.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningUtils.scala
Comment (HyukjinKwon) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningUtils.scala]: Sure.

File: docs/sql-programming-guide.md
Comment (HyukjinKwon) [inline docs/sql-programming-guide.md]: Doc shows as below:

**Pull request closed without merging**
